pkg/controller/deployment/deployment_controller.go

```diff
@@ -507,26 +496,85 @@ func (dc *DeploymentController) syncDeploymentStatus(allRCs []*api.ReplicationCo
+		if err = dc.updateDeploymentRevision(deployment, newRC.Annotations[deploymentutil.RevisionAnnotation]); err != nil {
+			glog.V(4).Infof("Error: %v. Unable to update deployment revision, will retry later.", err)
+		}
+	}
+	return newRC, oldRCs, nil
+}
+
+func revision(rc *api.ReplicationController) (int, error) {
+	v, ok := rc.Annotations[deploymentutil.RevisionAnnotation]
+	if !ok {
+		return 0, nil
+	}
+	return strconv.Atoi(v)
+}
+
+func maxRevision(allRCs []*api.ReplicationController) int {
+	max := 0
+	for _, rc := range allRCs {
+		if v, err := revision(rc); err != nil {
+			// Skip the RCs when it failed to parse their revision information
+			glog.V(4).Infof("Error: %v. Couldn't parse revision for rc %#v, deployment controller will skip it when reconciling revisions.", err, rc)
+		} else if v > max {
+			max = v
+		}
+	}
+	return max
+}
+
+func (dc *DeploymentController) getOldRCs(deployment extensions.Deployment) ([]*api.ReplicationController, []*api.ReplicationController, error) {
 	return deploymentutil.GetOldRCsFromLists(deployment, dc.client,
 		func(namespace string, options api.ListOptions) (*api.PodList, error) {
 			podList, err := dc.podStore.Pods(namespace).List(labels.SelectorFromSet(deployment.Spec.Selector))
 			return &podList, err
 		},
 		func(namespace string, options api.ListOptions) ([]api.ReplicationController, error) {
-			return dc.rcStore.List()
+			return dc.rcStore.ReplicationControllers(namespace).List()
 		})
 }
 
 // Returns an RC that matches the intent of the given deployment.
 // It creates a new RC if required.
-func (dc *DeploymentController) getNewRC(deployment extensions.Deployment) (*api.ReplicationController, error) {
+// The revision of the new RC will be updated to maxOldRevision + 1
+func (dc *DeploymentController) getNewRC(deployment extensions.Deployment, maxOldRevision int) (*api.ReplicationController, error) {
+	// Calculate revision number for this new RC
+	newRevision := strconv.Itoa(maxOldRevision + 1)
+
 	existingNewRC, err := deploymentutil.GetNewRCFromList(deployment, dc.client,
 		func(namespace string, options api.ListOptions) ([]api.ReplicationController, error) {
-			return dc.rcStore.List()
+			return dc.rcStore.ReplicationControllers(namespace).List()
 		})
-	if err != nil || existingNewRC != nil {
-		return existingNewRC, err
+	if err != nil {
+		return nil, err
+	} else if existingNewRC != nil {
+		if existingNewRC.Annotations == nil {
+			existingNewRC.Annotations = make(map[string]string)
+		}
+		if existingNewRC.Annotations[deploymentutil.RevisionAnnotation] != newRevision {
+			existingNewRC.Annotations[deploymentutil.RevisionAnnotation] = newRevision
+			glog.V(4).Infof("update existingNewRC's revision to %s - %+v\n", newRevision, existingNewRC)
+			return dc.client.ReplicationControllers(deployment.ObjectMeta.Namespace).Update(existingNewRC)
```

Comment:
The RC's revision annotation is updated before it get scaled. If there's update conflict it'll retry later before scaling it.